### PR TITLE
feat: add new command_runner for dsl cmd.rb

### PR DESCRIPTION
**Pull request by mathiusj**
### TL;DR

Closes: https://github.com/Shopify/roast/issues/455

Refactored DSL command execution with a new `CommandRunner` class that provides better process management, timeout support, and streaming output handling.

### What changed?

**New Files:**
- `lib/roast/dsl/command_runner.rb` - DSL-specific command runner (~179 lines)
- `test/roast/dsl/command_runner_test.rb` - Comprehensive test suite (15 tests)

**Modified Files:**
- `lib/roast/dsl/cogs/cmd.rb` - Simplified from 38 lines → 22 lines in execute method (42% reduction)
- `test/dsl/functional/roast_dsl_examples_test.rb` - Fixed regex to handle macOS extended attributes

**Key Features:**
- ✅ Callback-based stream handling (line-by-line)
- ✅ Timeout support with graceful cleanup
- ✅ **Process tracking to prevent orphaned processes** (teammate's critical requirement)
- ✅ `cleanup_all_children` for signal handling
- ✅ Handler exception safety (failures don't break command execution)
- ✅ Preserves `raw_output?` config for output stripping
- ✅ Zero impact on YAML workflows

### Test Results

```bash
# CommandRunner unit tests
bundle exec ruby -Itest test/roast/dsl/command_runner_test.rb
# ✅ 15/15 tests passing

# DSL functional tests
bundle exec ruby -Itest test/dsl/functional/roast_dsl_examples_test.rb
# ✅ 7/7 tests passing
```

**Manual DSL Workflow Testing:**
```bash
bundle exec ruby -e "require_relative 'lib/roast'; Roast::DSL::Workflow.from_file('dsl/prototype.rb')"
bundle exec ruby -e "require_relative 'lib/roast'; Roast::DSL::Workflow.from_file('dsl/step_communication.rb')"
```

### API Example

```ruby
# Basic usage (capture only)
stdout, stderr, status = CommandRunner.execute("echo", "hello")

# With streaming callbacks
CommandRunner.execute(
  "ls", "-la",
  stdout_handler: ->(line) { puts "[OUT] #{line}" },
  stderr_handler: ->(line) { puts "[ERR] #{line}" }
)

# With timeout
CommandRunner.execute("sleep", "30", timeout: 5)  # Raises TimeoutError after 5s
```

### Why make this change?

**Problems with old implementation:**
- 38 lines of duplicated threading logic in `cmd.rb`
- No process tracking (risk of orphaned processes)
- No timeout support
- Limited error handling for IO operations

**Benefits of new implementation:**
- 42% reduction in `cmd.rb` complexity
- Prevents orphaned processes via tracking
- Testable in isolation (15 unit tests)
- Reusable by other DSL cogs if needed
- Clean, focused API

### Notes

- **Sorbet warnings**: 2 expected splat-related warnings (matches existing CmdRunner pattern)
  - `lib/roast/dsl/command_runner.rb:46` - Open3.popen3 splat
  - `lib/roast/dsl/cogs/cmd.rb:165` - CommandRunner.execute splat
- **macOS compatibility**: Fixed test regex to handle extended attributes (`@` symbol in `ls -l` output)

### CODE REVIEW

**Comment by mathiusj** [comment]
* **#485** <a href="https://app.graphite.dev/github/pr/Shopify/roast/485?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Shopify/roast/485?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: test/roast/dsl/command_runner_test.rb [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+          "-c",
+          "echo 'line1' && echo 'line2' && echo 'line3'",
+          stdout_handler: ->(line) { lines << line },
+        )
+
+        assert_equal ["line1\n", "line2\n", "line3\n"], lines
+        assert_equal "line1\nline2\nline3\n", stdout
+      end
+
+      test "handlers still work with non-zero exit status" do
```
Comment (juniper-shopify) [inline test/roast/dsl/command_runner_test.rb]: we should also test that the captured output/error values are fully present for non-zero exit status too. There's a test for the exit status above, but it ignores the stdout/stderr return values

File: test/roast/dsl/command_runner_test.rb [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        stdout, stderr, status = CommandRunner.execute("echo", "hello")
+
+        assert_equal "hello\n", stdout
+        assert_equal "", stderr
+        assert_equal 0, status.exitstatus
+      end
+
+      test "captures both stdout and stderr separately" do
+        stdout, stderr, status = CommandRunner.execute(
+          "bash", "-c", "echo 'to stdout' && echo 'to stderr' >&2"
```
[the remark this answers is not in the document]
Comment (dersam) [reply test/roast/dsl/command_runner_test.rb]: Agreed, you shouldn't have to think about the semantics of "can I use pipes". We already let you chain commands like a terminal by default in legacy Roast, and that's a good feature that we should continue to support.

File: lib/roast/dsl/cogs/cmd.rb [hunk 3]
```diff
@@ -157,43 +157,26 @@ def raw_output?
...
-            # Wait for threads to finish
-            stdout_thread.join
-            stderr_thread.join
-
-            command_output = command_output.strip unless config.raw_output?
-            command_error = command_error.strip unless config.raw_output?
-
-            [command_output, command_error, wait_thread.value]
-          end #: as [String, String, Process::Status]
+          # Convert config booleans to handlers for CommandRunner
```
[the remark this answers is not in the document]
Comment (juniper-shopify) [reply lib/roast/dsl/cogs/cmd.rb]: Yeah... simple advice: do not look at *anything* outside of the `dsl/` section for style guidance. The rest of the app is mostly a vibe-coded mess that we've patched up here and there. It's why we're incubating this new architecture and planning a hard cut-over

File: lib/roast/dsl/cogs/cmd.rb [hunk 4]
```diff
@@ -157,43 +157,26 @@ def raw_output?
...
+          stdout_handler = config.print_stdout? ? ->(line) { $stdout.print(line) } : nil
+          stderr_handler = config.print_stderr? ? ->(line) { $stderr.print(line) } : nil
+
+          stdout, stderr, status = CommandRunner #: as untyped
+            .execute(
+              input.command,
+              *input.args,
```
[the remark this answers is not in the document]
Comment (mathiusj) [reply lib/roast/dsl/cogs/cmd.rb]: Hey @juniper-shopify! I actually tried that approach first but hit a Sorbet constraint:

```
lib/roast/dsl/command_runner.rb:XX: Splats are only supported where the
size of the array is known statically https://srb.help/7019
```

Even with `args = args #: as untyped` (not that we would want that anyway 😅) , Sorbet still complained about the splat at the `Open3.popen3` call because it can't verify the array size at compile time.

Sorbet accepts the current solution because the splat is part of the method signature from the start rather than dynamically splatting an array.

Is there a way to use an Array parameter that I've missed? Or is this a known Sorbet limitation with dynamic splats?

Happy to try a different approach if there's a way around the Sorbet constraint!

File: lib/roast/dsl/command_runner.rb [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        rescue Timeout::Error
+          cleanup_child_process(pid) if pid
+          raise TimeoutError, "Command timed out after #{timeout} seconds"
+        end
+
+        # Cleanup all tracked child processes
+        #
+        # This is called on shutdown/signal handling to ensure no orphaned processes
+        #: -> void
+        def cleanup_all_children
```
[the remark this answers is not in the document]
Comment (mathiusj) [reply lib/roast/dsl/command_runner.rb]: After reviewing your comments and exploring this in the codebase, here is what I believe we need for this PR and why:

**1. OS does handle Ctrl+C cleanup automatically for direct execution**

```
roast → ls
  ↓ Ctrl+C
OS kills both ✅
```

**2. BUT we need cleanup for timeouts:**

```ruby
CommandRunner.execute("sleep", "100", timeout: 0.1)
# After 0.1s, Timeout::Error is raised
# Without cleanup, sleep keeps running! ⚠️
```

**What I implemented (minimal for timeout handling):**
- Track PID during execution
- On timeout: kill process with TERM, then KILL if needed
- Helper methods: `kill_process`, `process_running?`
- Test to verify process is actually killed

**What we do not need / I removed from this PR (deferred to shell support):**
- Hash tracking, mutex, `cleanup_all_children`
- Signal handlers
- Process group management

**The grandchildren orphaning you're thinking about only matters with shell execution:**

```
# Shell execution (future feature)
roast → sh -c "sleep 100 | grep foo"
       ↓
       sh → sleep, grep (grandchildren)
  ↓ Ctrl+C kills roast and sh
But sleep + grep become orphaned! ⚠️
```

When we add shell command support, we can implement proper process group handling:
```ruby
Open3.popen3(*args, pgroup: true) do |..., wait_thread|
  pgid = Process.getpgid(wait_thread.pid)
  Process.kill("-TERM", pgid)  # Kills entire process group
end
```

Current PR has minimal cleanup for timeout correctness only now. Let me know if that sounds good to you 🙏 I'll push a PR to the stack sometime early next week for shell command support. I [created an issue](https://github.com/Shopify/roast/issues/489) to track that follow-up work.

File: test/roast/dsl/command_runner_test.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (juniper-shopify) [reply test/roast/dsl/command_runner_test.rb]: Yes, just passing the whole thing as a string to Open3 does work with pipes and such. Ultimately, we may just want to simplify the inputs for the `cmd` cog,  to drop the `args` array and just take a string.

(To that end, I do want to add some functionality to it to let you configure environment variables, clear the environment, choose the shell to run the command with, etc. so we'll need to do some orchestration under the hood one way or the other)

**Comment by nfgrep** [comment]
>While I'd like to avoid a generic utils/ or helpers/ structure, like we discussed on Slack earlier, I also want to avoid a ton of classes piling up in the 'root' directory of the source hierarchy (taking lib/roast/dsl as effectively the main source root). Let's all take this as an opportunity to agree on the structure we want to use for stuff like this now.

I'd prefer piling up in root over an organizational symbol as generic as "util" or "helper". If there are many things in root then there's a nice forcing function for us to find commonalities and group them into sub-folders, if we act to early on that we get sub-folders like "util", "tools", "helpers" etc. 

I think the ideal future state is that things like `CommandRunner`, logging, etc all live at true root (where we can eventually identify groups and move into subfolders etc) and the DSL-specific stuff is in what we currently consider the "root" of this new codepath.

Right now we're overloading `lib/roast/dsl` to both mean "the root of the new thing" and "DSL-specific implementation details". We could rename our "root" to be something like `lib/roast/neo`, then all our generic stuff like CommandRunner could live there, and we could have a `lib/roast/neo/dsl` that is just DSL-specific files like all the context management stuff, etc.

**Comment by juniper-shopify** [comment]
@mathiusj I just pushed up an update to your PR with a 'fix' for the test that should have been failing (see https://app.graphite.dev/github/pr/Shopify/roast/485/feat-add-new-command_runner-for-dsl-cmd.rb#comment-PRRC_kwDOOW6Stc6TtxjE). I'd already implemented this change when trying to figure out why the test suite was so slow on my local machine, so I figured I'd just push it up instead of commenting it as a suggestion, but feel free to change if you have a better way in mind to run the test.

I have not investigated what the root cause is; I just know it doesn't actually time out 😢

File: test/roast/dsl/command_runner_test.rb [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        sleep(0.15)
+
+        # The sleep process should be dead
+        # We can't easily get the PID from outside, but we can verify
+        # that a very short timeout doesn't leave sleep running
+        # by checking process list (external check, not using CommandRunner)
+        # rubocop:disable Roast/UseCmdRunner
+        output = %x(ps aux | grep "sleep 100" | grep -v grep)
+        # rubocop:enable Roast/UseCmdRunner
+        assert_empty output, "sleep process should be killed after timeout"
```
[the remark this answers is not in the document]
Comment (mathiusj) [reply test/roast/dsl/command_runner_test.rb]: thanks for spotting this! I'll fix it ASAP 🙏

**Comment by juniper-shopify** [comment]
### Merge activity

* **Oct 31, 2:48 AM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/Shopify/roast/485).
* **Oct 31, 2:48 AM UTC**: @juniper-shopify merged this pull request with [Graphite](https://app.graphite.dev/github/pr/Shopify/roast/485).
